clients/src/main/java/org/apache/kafka/clients/admin/AlterConsumerGroupOffsetsResult.java

```diff
@@ -0,0 +1,88 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.admin;
+
+import java.util.Map;
+
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.KafkaFuture.BaseFunction;
+import org.apache.kafka.common.KafkaFuture.BiConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.internals.KafkaFutureImpl;
+import org.apache.kafka.common.protocol.Errors;
+
+/**
+ * The result of the {@link AdminClient#alterConsumerGroupOffsets(String, Map)} call.
+ *
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class AlterConsumerGroupOffsetsResult {
+
+    private final KafkaFuture<Map<TopicPartition, Errors>> future;
+
+    public AlterConsumerGroupOffsetsResult(KafkaFuture<Map<TopicPartition, Errors>> future) {
+        this.future = future;
+    }
+
+    /**
+     * Return a future which can be used to check the result for a given partition.
+     */
+    public KafkaFuture<Void> partitionResult(final TopicPartition partition) {
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+
+        this.future.whenComplete(new BiConsumer<Map<TopicPartition, Errors>, Throwable>() {
+            @Override
+            public void accept(final Map<TopicPartition, Errors> topicPartitions, final Throwable throwable) {
+                if (throwable != null) {
+                    result.completeExceptionally(throwable);
+                } else if (!topicPartitions.containsKey(partition)) {
+                    result.completeExceptionally(new IllegalArgumentException(
+                        "Alter offset for partition \"" + partition + "\" was not attempted"));
+                } else {
+                    final Errors error = topicPartitions.get(partition);
+                    if (error == Errors.NONE) {
+                        result.complete(null);
+                    } else {
+                        result.completeExceptionally(error.exception());
+                    }
+                }
+
+            }
+        });
+
+        return result;
+    }
+
+    /**
+     * Return a future which succeeds if all the alter offsets succeed.
+     */
+    public KafkaFuture<Void> all() {
```

Comment:
That's a fair point but I am not sure what the best one is. The rational behind not looking at individual topic/partitions was that it allows to use `all()` to wait for the completion of the request and then check the individual results. In this case `all()` fails only if the whole group has failed.

To be more concrete, it allows to do the following:

```java
DeleteConsumerGroupOffsetsResult result = ...;

try {
  // wait for the whole group, only raise when a group level or
  // transport level exception affection the whole request occurs
  result.all().get()

  // inspect individual topic/partition
  try {
    result.partitionResult(...).get()
  } catch (Exception e) {
    // handle partition exception
  }
} catch (Exception e) {
  // handle group level exception
}
```

I think that this facilitates the error handling. What do you think?

Reply:
That's an interesting point. I think the usual semantics of `all` is to only succeed if all individual operations have succeeded. It's sort of designed for lazy error handling I guess. If users care about the individual operations, they can check them individually. Otherwise they have a convenient way to check for any errors. Based on what I've seen, this tends to be the most frequent use. I think also part of the idea is to abstract away from the underlying requests. Some of the admin APIs result in multiple broker requests which makes exposing the full granularity of errors quite cumbersome.

Reply:
Just made a pass on all XXXResult classes and I think the API semantics are a bit inconsistency in general: originally I thought we only need the `all` function if the result contains futures in the form of `Map<..., KafkaFuture<...>>` which potentially requires one trip for each nested future, and the `all` function is used as a lazy way to check that all entries have completed successfully. But some (e.g. `RemoveMemberFromGroupResult` in form of `Map<MemberIdentity, KafkaFuture<Void>>`) actually only requires one request too, so all futures would actually be always completed at the same time. For those cases we do not need an `all` function either.

But it seems like for results that only contain a `KafkaFuture<Object>` we also have a dummy `all` function, and many of their `all` semantics are different too.

Honestly I think not all `results` needs an `all` function, but it seems we are already a bit messy here..

Reply:
Yeah, unfortunately the admin APIs have such a big surface area it's hard to maintain consistency. I think the original intent is what I described though.